### ISSUE TITLE
Finalize gettext implementation for ZikulaTwig

### ISSUE THREAD

**Issue by craigh** [zikula/core#2260]
There are currently two options for how Zikula will implement get text translation in Twig templates:

  1. so called "native" implementation (http://twig.sensiolabs.org/doc/extensions/i18n.html) which is currently ***enabled*** in the core as of Core 1.4.0 but unused
  2. The zikula method which is implemented as a [Twig Extension of the core bundle](https://github.com/zikula/core/blob/1.4/src/lib/Zikula/Bundle/CoreBundle/Twig/Extension/GettextExtension.php)

There may be advantages to either, but questions remain about the value and implementation of both.

  1. does the native implementation allow for multiple translation domains?
  2. how are strings extracted using the zikula twig extension?

I'm sure there are others.

I think the zikula method appears simpler:

native:
```
{% trans %}
    Hey {{ name }}, I have one apple.
{% plural apple_count %}
    Hey {{ name }}, I have {{ count }} apples.
{% notes %}
    This is shown in the user menu. This string should be shorter than 30 chars
{% endtrans %}
```

zikula:
`{{ _fn(/*!This is shown in the user menu. This string should be shorter than 30 chars*/'Hey %1$s, I have one apple', 'Hey %1$s, I have %2$s apples', array($name, $count), $count) }}`

the new ZikulaInstallerBundle already [implements](https://github.com/zikula/core/tree/1.4/src/lib/Zikula/Bundle/CoreInstallerBundle/Resources/views) usage of the 'zikula' method but currently, the strings are not extracted using the Gettext module because the regex is not compatible. (refs https://github.com/zikula-modules/Gettext/issues/8 and https://github.com/zikula-modules/Gettext/pull/9)

refs: 
 - https://github.com/zikula/core/issues/1753
 - https://github.com/zikula/core/issues/2191

ping @cmfcmf @Guite @drak

**Comment by craigh** [comment]
ah - here is another concern for the native method:
> The i18n extension only works if the PHP gettext extension is enabled.

**Comment by rallek** [comment]
To have this labeled for Milestone 2.0.0 makes the 1.4 branch useless for me :-(

I can not accept english sentences on a total german page.

**Comment by Guite** [comment]
@rallek : afaik the only place where Twig is currently used is the installer/upgrader. So it won't affect your actual site.

**Comment by rallek** [comment]
ok. I did not know this. Than I am fine with 2.0.0. Sorry for claiming

**Comment by Kaik** [comment]
At the moment for translating we are using tools in lib/i18n
ZLanguage where all the paths are solved
ZGettext is for reading 
In old modules old abstractbase translation is loaded here https://github.com/zikula/core/blob/1.4/src/lib/legacy/Zikula/AbstractBase.php#L149
And all the stuff works 

For twig and new module style we have things here https://github.com/zikula/core/blob/1.4/src/lib/Zikula/Core/AbstractBundle.php#L73
This I believe are meant to be working with this
https://github.com/zikula/core/tree/1.4/src/lib/Zikula/Common/I18n
and this https://github.com/zikula/core/blob/1.4/src/lib/Zikula/Core/Controller/AbstractController.php#L23
So it will be symfony translator wise "whichever we would use our gettex implementation or nativ trans" with some work toward zikula gettext is done already.

Now we can use {{ __('string') }} in twig using AbstractController and it works for zikula domain because it's binding is hardcoded as default both "domain binding" and domain "detection".

To change this for module we need to 
bind module domain - set paths for ZGettext::reader and this is not yet done for new modules we can do this in controller for now ( its obviously a hack)  
detect domain in "Twig Extension of the core bundle" so we don't have to specify domain like {{ __('string', 'domain') }} and use it like this {{  __('string') }} 

We can stick to our gettext implementation 'life as we know it' but this should be done in a "TranslatorAwareInterface" manner whatever that mean.
I still do not know where this thing should happen https://github.com/zikula/core/blob/1.4/src/lib/i18n/ZLanguage.php#L421
I new way might be ommiting ZLanguage file and work directly with ZGettextFunctions.php

**Comment by Kaik** [comment]
I think we can have both native symfony and zikula gettext working together. We will use symfony nativ translator service and then do one of the options here http://stackoverflow.com/questions/15550002/advanced-customization-of-translations-in-symfony2 
We can extend it by our gettext __() ... Things are already done but there are still using our lib/i18n while I think it should be solved inside https://github.com/zikula/core/tree/1.4/src/lib/Zikula/Common/I18n without need for ZLanguage ZGettext at all.

Translator is adding new tag {% trans %} to Twig
We are using this for our tags
https://github.com/zikula/core/blob/1.4/src/lib/Zikula/Bundle/CoreBundle/Twig/Extension/GettextExtension.php 
We may be able I do not know yet to extend translator twig extension by our zikula gettext extension so we will have all the gettext stuff in one place.

**Comment by Kaik** [comment]
Hi, yesterday I was testing symfony nativ translator. 
I have extended Translator and PoFileLoader (all placed in Zikula/Common/Translator) with custom translator class in config.xml
Then Used it like this (currently zikula breakit module NewController)
```
        $translator = $container->get('translator');
        
        $translator->addLoader('po', new PoFileLoader());
        //$translator->addResource('po',$this->getLocalePath().'/' . $translator->locale. '/LC_MESSAGES/'.$this->getTranslationDomain().'.po' , $translator->locale , $this->getTranslationDomain() );
        $translator->addResource('po','path to zikula.po' , 'pl' , 'zikula' );
         
        				
        $catalogue = $translator->getCatalogue('pl');
        $translated = $translator->trans(
	    'Default',
	    array(),
	    'zikula',
	    'pl'
		);
```
Twig
```
            {% trans from "zikula" %}Error! Cannot determine valid 'cid' for edit mode in 'ZikulaCategoriesModule_admin_edit'.{% endtrans %}
            {% trans %}Error! Cannot determine valid 'cid' for edit mode in 'ZikulaCategoriesModule_admin_edit'.{% endtrans %}
            {{ dump(translated) }}
            {{ dump(translator) }}
            {{ dump(catalogue) }}
```
And results are 
![screenshot from 2015-05-16 22 58 58](https://cloud.githubusercontent.com/assets/966910/7669620/b44b12a6-fc7d-11e4-8051-085316df39b3.png)

That's what I have at the moment. Now I need, to move this add resource stuff somewhere? AbstractController?(there is already another translator zikula gettext implementation but I want to omit that, and base everything on sf translator service ) AbstractBundle? If you have any ideas please let me know. Actually I tried add resource in controller constructor but it needs container and this is null or something so I'm lost :(

**Comment by craigh** [comment]
Hi @Kaik ,

I am thankful that you are working on this, but I truly have no idea about any of it. I hope @Guite or @cmfcmf (or anyone) (@shefik) has some idea, as I cannot be of any help on this one. sorry!

You might Try the new Pages module (you will need my un-merged Core PR first) as it has been completely refactored to twig also.

Also, I had a question as I was working on twig stuff. In some places I see `{{ p.__('My string') }}` and in others only `{{ __('My string') }}`. I would like to know what the `p` does in that case.

**Comment by shefik** [comment]
I'll take a look at this soon.

**Comment by Kaik** [comment]
Hi, actually I do not know where to start, so as far as I understand it and can explain.
Currently we can do __('string') in php , {gt text="string"} smarty , {{ __('My string') }} twig ,  {{ p.__('My string') }} twig. All of them use /lib/i18n. I18n is working together with /lib/StreamReader/ this is I believe, why we do not need servers with gettext extensions enabled(might be wrong on this). 
__() and other in php, are global function's loaded here https://github.com/zikula/core/blob/1.4/composer.json#L27 lets forget this for a second.
Gettext translations .pot files not interested at the moment .po this are translations as object, human and machine readable .mo compiled translations machine readable.
Zikula have /Resources/locale folder Symfony /Resources/translations
Zikula have lang/LC_MESSAGES/domain.po/mo hierarchy Symfony /domain.lang.po/mo
Domain for Zikula is "module/theme" oriented (we have 'module_dizkus', 'theme_andreas08' domains legacy new bundle way @shefik changed it to 'dizkus') while for Symfony its more like navigation, validators etc...

The trick is to have "one place array object cached" where all needed translations are loaded from files and we can access them using locale and domain string. Zikula uses ZLanguage and ZGettext singeltons - one place. Symfony uses nice object thing. 
Zikula when any page is viewed load current lang translations for domains core, theme, module, plugins etc... into that singeltone and this is loaded when mentioned part is loaded AbstractBase, AbstractModule, AbstractController etc... all legacy stuff. Easy first loaded then retrieved when needed using __() etc... 
@craigh 
New Controller - here are two things, something that Drak was trying to do, it is some kind of new way of doing what legacy does, still using 'Zikula gettext as base ZLanguage ZGettext' and this is {{ p.('My string') }} and it is actually drak's Translator class as decoration, it has __() functions that use Zikula gettext.
Second thing Core Twig Gettext Extensions {{ __() }} do the same, they all eventually redirects to /lib/i18n/

All of the above have nothing to do with Symfony nativ Translator.
Simply what I'm trying to do now is to replace core of translations ZLanguage ZGettext with Symfony Translator. First load then retrieve, so first I need to load our .po files from our Zikula folder hierarchy into SF translator there is few ways of doing that. I wanted to use what is working now - loading from /Resources/translations and at the moment I 'succeed' using CoreBundle DependencyInjection and this does load our po files nicely from zikula and brake-it module and {% trans()  with %} do work. 
Translations are cached but I still do not know if this is the right place to load it.
 long one ;)

**Comment by Kaik** [comment]
Hi,
Some real examples. 
https://github.com/Kaik/core/tree/symfony_translator
This should load all domain.lang.po files from Resources/translations SF
This should load all domain.po files from Resources/locale/$lang/$catalogue/ ZK
$lang = pl etc.. $catalogue = LC_MESSAGES
You should be able to do {% trans from "zikula" %}Error! Cannot determine valid 'cid' for edit mode in 'ZikulaCategoriesModule_admin_edit'.{% endtrans %} in Twig.
Sources:
https://github.com/symfony/FrameworkBundle/tree/2.6
https://github.com/symfony/FrameworkBundle/blob/2.6/DependencyInjection/FrameworkExtension.php#L627
We may want to add our own Translator
https://github.com/symfony/FrameworkBundle/blob/2.6/Translation/Translator.php
Now I will try to detect domain so we can do {% trans %}Error! Cannot determine valid 'cid' for edit mode in 'ZikulaCategoriesModule_admin_edit'.{% endtrans %} without need to use from "zikula". When that succeed we can try to do BC so __() , {gt text="" } , {{  __() }}

**Comment by craigh** [comment]
I'm not sure I would qualify `{{ __('my string') }}` as BC since it never existed. (it is also my preference for how to handle this).

I don't think we need to worry about `{gt text="my string"}` because that is smarty syntax anyway (it would have to be `{% gt string="my string" %}` but the first is better than this anyway).

**Comment by Kaik** [comment]
Yes it is not BC actually (if everything works as I expected) there might by only one thing to consider using {{  }} notation Why SF devs used {%  %} for translations?
From wiki
Twig defines three kinds of delimiters:
{% ... %}, which is used to execute statements, such as for-loops.
{{ ... }}, which is used to print the content of variables or the result of evaluating an expression.
{# ... #}, which is used to add comments in the templates. These comments aren't included in the rendered page.
I just started to prefer {% trans %} ;) but {{ __() }} is standard I want to achieve.

> I don't think we need to worry about {gt text="my string"} because that is smarty syntax anyway (it > would have to be {% gt string="my string" %} but the first is better than this anyway).

BC will not change anything in usage it will only replace the translations engine from ZGettext to SF translator, at least this is the concept.

I forget to mention that you need to clear cache every time you add translation file or change name etc even in dev env.

**Comment by craigh** [comment]
> there might by only one thing to consider using {{ }} notation

the basic idea is that `{{ }}` delimiters are used to define functions that output something and are defined in the code as a callable. `{% %}` delimiters are used typically for control structures and other items. These are harder to define/create in the code. If the function is already created, then it doesn't matter since the hard work is already done. As far as I know, they are not interchangeable.

**Comment by Kaik** [comment]
Hi,
I have just pushed to https://github.com/Kaik/core/tree/symfony_translator update to Core Twig Extensions so now it is possible to do:
Simple translations, for current domain and for 'zikula' domain passed as argument
```
	{{ __('Done!') }}
	{{ __('Done!', 'zikula') }}
```
With variable replacement, for current domain and for 'zikula' domain passed as argument
```
	{{ __f('Done! Saved the %s category.',{'%s':'test'}) }}
	{{ __f('Done! Saved the %s category.',{'%s':'test'}, 'zikula') }}
```
Let me know what you think. This is obviously an idea how it can work.

**Comment by craigh** [comment]
> Let me know what you think.

cool! :+1:

**Comment by craigh** [comment]
closed by #2460
